docs/man/mcloc.1
.de hL
\fI\ \ \\$1\l'|2\(ul'\fR
..
.de bL
\fI\\$1\ \\$2\l'|2\(ul'\fR
..
.AM
.Dd Jan 18, 2020
.Dt mcloc 1
.Os mcplex.net
.Sh NAME
.Nm mcloc
.Nd count lines of code in source files
.Sh SYNOPSIS
.Nm
.Op Fl C
.Op Fl c Ar config
.Op Fl E Ar extension(s)
.Op Fl L Ar language(s)
.Op Fl S Ar scanner(s)
.Op Fl d | e | l | s
.Op Fl g
.Op Fl j Ar numThreads
.Op Fl M Ar cocomo_config
.Op Fl m
.Op Fl o
.Op Fl q
.Op Fl a | r
.Cm path(s)
.Sh DESCRIPTION
.Xr \fBmcloc 1\fR
is a simple command line utility for obtaining a reasonably accurate
count of lines of code and comments in Ada, Assembly, awk,
bison, C, C++, C#, Clojure, Cmake, CSS, D, Dart, Erlang,
Bflex, Fortran, Go, Groovy, Haskell, HTML, IDL, java, JSON,
Julia, Kotlin, Lisp, Lua, m4, make, man, Objective-C, Perl, PHP,
Prolog, protobuf, Python, QML, R, Ruby, Rust, Scala, Scheme, sed, SQL,
Swift, TeX/LaTeX, TOML, TypeScript, YAML and shell script source
files.
.Pp
The count is not always 100% accurate.  For 100% accuracy, we'd have
to fully parse the source code, which is a non-trivial activity
for the languages I am most interested in (C++, for example).  The
point of
.Xr \fBmcloc 1\fR
is quick assessments of source code size, mainly to give a developer a
sense of the effort required to maintain a set of source code files.
It's particularly useful in refactoring efforts, where reduction of
source code size may be a stated objective.
.Ss Optional arguments
The following optional command line arguments are available:
.Pp
.Bl -tag -width indent
.It Fl C
Print the configuration and exit (do not scan files).  Note this will
incorporate any changes made with the \fI-E\fR and \fI-S\fR arguments.
.It Fl c Ar config
Use \fIconfig\fR as the configuration file instead of the default
configuration file.
.It Fl E Ar extension(s)
Only scan files with an extension specified in \fIextensions(s)\fR,
which is a comma-separated list of file extensions
(e.g. \fI.lex,.c,.h\fR).
.It Fl L Ar language(s)
Only scan files for the languages specified in \fIlanguage(s)\fR,
which is a comma-separated list of languages (e.g. \fIc++,python\fR).
.It Fl S Ar scanners(s)
Only activate the scanners named in the comma-separated list of
\fIscanners\fR, from the following list:
.Pp
Ada, CSS, Clojure, Cpp, D, Erlang, Flex, Fortran, Haskell, HTML, Julia,
Kotlin, Lisp, Lua, m4, Manpage, NoComments, Perl, Prolog, Python, Ruby,
Rust, Shell, SQL, TeX
.It Fl a
Sort output entries by lines of code, in ascending order.
.It Fl d
Print counters for each directory (do not show files).
.It Fl e
Print counters for each file extension instead of counters for each source
file and/or directory.
.It Fl l
Print counters for each language instead of counters for each source file
and/or directory.
.It Fl s
Print counters for each scanner instead of counters for each source file
and/or directory.
.It Fl g
Include files generated by bison, flex, moc, rcc or uic.
.It Fl j Ar numThreads
Specify the number of threads to use to parse source files.  Note there
are tradeoffs here, and more threads may or may not be faster than fewer
threads.
.Pp
The default thread count for parsing source files is the concurrency
reported by std::thread::hardware_concurrency().
.It Fl M Ar cocomo_config
Use \fIcocomo_config\fR as the COCOMO project configuration file.
.It Fl m
Report COCOMO Intermediate estimates.
.It Fl o
Do not recurse into subdirectories.
.It Fl q
Only print totals (quiet).
.It Fl r
Sort output entries by lines of code, in descending order.
.El
.Ss Required argument(s)
.Xr \fBmcloc 1\fR requires one or more \fBpath(s)\fR after the optional
arguments.  Each of these arguments must be a path to a file or directory.
.Ss Output
The output of
.Xr \fBmcloc 1\fR
has 4 columns.  In the first line (column titles) and the last line
(TOTAL), the first column is intentially left blank to make it easy
to grep for only data lines in the output.  For all other lines in the
output, the first column contains a single-character indicator of the
type of the row:
.Pp
.Bl -tag -width "D" -compact -offset indent
.It D
directory
.It F
file
.It G
generated file
.It E
filename extension
.It L
language
.It S
scanner
.El
.Pp
The second column of a data row contains the lines of code counter.
.Pp
The third column of a data row contains the comment counter.
.Pp
The fourth column contains the row's label.  This may be a path to
a directory, a filename, a file extension, a language, a scanner, or
\fBTOTAL\fR.
.Pp
See
.Sx EXAMPLES
for example output.
.Sh ENVIRONMENT
The \fIMCLOC\fR environment variable can be used to set default options.
For example, a value of 'a:l' would enable ascending sort ('-a') and
print-by-language ('l') options.  A value of 'a:l:L=c++,c' would do the
same and also only select C and C++ files for counting.  Command line
arguments override the settings in \fIMCLOC\fR environment variable.
.Sh NOTES

The main goal of
.Xr \fBmcloc 1\fR
is fast, reasonably accurate counting of lines of code and comments in
source files.  I primarily use it on C and C++ projects (Which may
include flex, bison, and Qt moc/uic/rcc files).  I can't really vouch
for the accuracy of counts for other languages simply because I
haven't used
.Xr \fBmcloc 1\fR for any significant code bases in other languages.
.Pp
.Xr \fBmcloc 1\fR
occupies an area somewhere closer to high speed than complete accuracy,
by design.  For my personal source code style, I've found it to produce
reliable lines-of-code metrics, and it's always done so at a rate that
makes me happy to use it.  Your mileage may vary, as I wrote this tool
for my own use.
.Pp
Below is a table showing lines/second while processing C++ on a few
machines in my home.
.Pp
.TS
box tab(:);
lb|lb|lb|lb
l|l|l|r.
CPU:I/O:O/S:lines/sec
_
Threadripper 3960X:NVMe (PCI 4 x4):Ubuntu 20.04:75,000,000+
Core i9 2.9GHz:NVMe (PCI 3 x4):MacOS Mojave:14,000,000+
E3-1270 V2 3.50GHz:SATA SSD:FreeBSD 12.2:20,000,000+
2x Xeon L5640 2.27GHz:SATA SSD:FreeBSD 12.2:25,000,000+
Core i5 2405S 2.50GHz:SATA SSD:FreeBSD 12.2:10,000,000+
ARM Cortex-A72 1.5GHz:microSD:Raspbian 10:600,000 
.TE
.br
\X'ps:'\c
.br
.Pp
I find these numbers sufficient for my own use.  Much of the time I'm
looking at a smaller code base or a smaller piece of a larger code
base.  Usually less than 200,000 lines of code.
.Pp
.Sh EXAMPLES
Note that \fBmcloc\fR will use locale formatting for line counts.
Some of the examples below show output for the C locale, while others
shouw output for en_US.UTF-8.

An example from the classes/src directory of
.Xr \fBmcloc 1\fR
itself:
.Pp
.Bd -literal -offset 0
% mcloc *Config*
.hL "    CODE COMMENTS PATH                                               "
D      560      220 .
F      260      138   DwmMclocConfig.cc
F       50       39   DwmMclocConfigLex.lex
.bL "F" "     250       43   DwmMclocConfigParse.y                             "
\fB\ \ \ \ \ \ \ 560      220 TOTAL\fR  (1.129ms, 690.9K lines/s, 0% sys 434% usr)
.Ed
.Pp
What if I include generated source files and sort by size?  Note how
in this case, the generated code (which we never modify) is much more
code than what we actually maintain.  This is a case where it's
clearly misleading to include generated code in a lines-of-code
metric.
.Pp
.Bd -literal -offset 0
% mcloc -a -g *Config*
.hL "    CODE COMMENTS PATH                                               "
D     3638      905 .
F       50       39   DwmMclocConfigLex.lex
G       56       33   DwmMclocConfigParse.hh (DwmMclocConfigParse.y)
F      250       43   DwmMclocConfigParse.y                              
F      260      138   DwmMclocConfig.cc
G     1451      411   DwmMclocConfigLex.cc (DwmMclocConfigLex.lex)
G     1571      241   DwmMclocConfigParse.cc (DwmMclocConfigParse.y)
\fB\ \ \ \ \ \ 3638      905 TOTAL\fR  ( 887us, 5.122M lines/s, 0% sys 394% usr)
.Ed
.Pp
What about per file extension, sorted by lines of code?
.Pp
.Bd -literal -offset 0
% mcloc -ae .
.hL "       CODE    COMMENTS EXTENSION/EXPRESSION                             "
E          79           2 "Makefile"
E         398         109 .y
E        1398         601 .cc
.bL "E" "       1408        1246 .lex                                                 "
\fB\ \ \ \ \ \ \ \ \ 3283        1958 TOTAL\fR  (4.426ms, 1.184M lines/s, 126% sys 642% usr)
.Ed
.Pp
How many lines of code per language in libDwm, sorted by lines of code?
.Pp
.Bd -literal -offset 0
% mcloc -al libDwm/trunk
.hL "       CODE    COMMENTS LANGUAGE                                         "
L         217           8 CSS
L         323           0 man
L         439         111 Shell
L         593          19 Make
L         698          58 m4
L         880         308 flex
L       1,061         102 bison
.bL "L" "     34,307      22,730 C++                                              "
\fB\ \ \ \ \ \ \ 38,518      23,336 TOTAL\fR  (9.65ms, 6.41M lines/s, 219% sys 141% usr)
.Ed
.Pp
How many lines of code per language in qt5?  Note this is a bit misleading
since Qt uses '.h' for C++ header files, and hence we count many of Qt's
header files as C instead of C++.
.Pp
.Bd -literal -offset 0
% mcloc -a -l ~/gits/qt5
.hL "     CODE    COMMENTS LANGUAGE                                         "
L        52           3 SQL
L        87          25 sed
L       183          55 awk
L       222          51 Scheme
L       351          58 Swift
L       662         275 Ruby
L       884         282 YAML
L       929         175 Dart
L      1042         124 TeX
L      1116          47 D
L      7887        2804 Make
L      7973         663 Lua
L      8648        8330 PHP
L      8998           6 QtRc
L     10371         843 bison
L     13943        2514 Go
L     27503        5446 Cmake
L     29944       47914 IDL
L     40095         466 man
L     41457       19496 protobuf
L     42749         524 Lisp
L     53406       18938 Shell
L     54689        5720 CSS
L     57084         407 flex
L     57954       13193 C#
L     64504        2531 m4
L     71597          27 QtUi
L     84993       14855 Perl
L    129891       43134 Java
L    232899       42446 Objective-C
L    242735       19706 Assembly
L    335257      189304 QML
L    475669      187102 Python
L    488130        2039 TypeScript
L    531711       13136 HTML
L    765125           0 JSON
L    822643        9663 XML
L   1188976      943066 JavaScript
L   7853913     2566652 C
.bL "L" " 12892785     2270945 C++                                                  "
\fB\ \ \ 26649057     6432965 TOTAL\fR  (1.257s, 26.32M lines/s, 265% sys 586% usr)
.Ed
.Pp
We can get an idea of the naming conventions for Qt C and C++ header files
by listing by file extension and selecting particular file extensions:
.Bd -literal -offset 0
% mcloc -E.c,.cc,.cpp,.cxx,.h,.hh,.hxx -a -e ~/gits/qt5
.hL "      CODE    COMMENTS DIR                                             "
E       1293        1393 .hxx
E      33028        2207 .cxx
E      54910       14204 .hh
E    3428463      768742 .c
E    4378933     1942584 .h
E    5210036     1312353 .cpp
.bL "E" "   7091067      947043 .cc                                                  "
\fB\ \ \ \ 20197730     4988526 TOTAL\fR  (350.5ms, 71.86M lines/s, 414% sys 1165% usr)
.Ed
.Pp
How many lines of code in each directory in libDwm?
.Pp
.Bd -literal -offset 0
% mcloc -d libDwm/trunk
.hL "      CODE    COMMENTS DIR                                             "
D      1,191         169 libDwm/trunk/
D         16           0 libDwm/trunk/apps
D        112          44 libDwm/trunk/apps/dwmnet
D        121          60 libDwm/trunk/apps/dwmspath
D        256          56 libDwm/trunk/apps/dwmwhat
D        191          51 libDwm/trunk/apps/fbsddeps
D         47          43 libDwm/trunk/apps/fixcflags
D        276          54 libDwm/trunk/apps/mcping
D        537         111 libDwm/trunk/apps/mkfbsdmnfst
D        139          50 libDwm/trunk/apps/psg
D        239           8 libDwm/trunk/doc
D        176          13 libDwm/trunk/doc/examples
D      8,342      11,515 libDwm/trunk/include
D         12           0 libDwm/trunk/packaging
D     16,486       6,996 libDwm/trunk/src
.bL "D" "    10,377       4,166 /Users/dwm/svn/dwm/libDwm/trunk/tests      "
\fB\ \ \ \ \ \ 38,518      23,336 TOTAL\fR  (2.65ms, 23.4M lines/s, 455% sys 450% usr)
.Ed
.Sh FILES
.Pa /usr/local/etc/mcloc.cfg
is the global configuration file.
.Pp
.Pa ${HOME}/.mcloc
will be used instead of the global configuration file, if present.
.Pp
.Pa /usr/local/etc/mcloc_cocomo1_imtermediate.cfg
is the default COCOMO project configruation file.
.Sh SEE ALSO
.Xr \fBawk 1\fR ,
.Xr \fBbison 1\fR ,
.Xr \fBflex 1\fR ,
.Xr \fBperlsyn 1\fR ,
.Xr \fBsed 1\fR
.Sh HISTORY
I first wrote
.Xr \fBmcloc 1\fR
in 1998.  At the time, we didn't have a lot of tools to count lines of
C++ code, and none of them could count generator files (flex and bison
input) while ignoring generated files (flex and bison output).
.Xr \fBmcloc 1\fR
was born when I needed to keep track of my progress on various
flex/bison lexers/parsers (SMIv2 MIB compiler, BGP route log parser,
etc.).  It's fairly typical for flex and bison output to be two
decimal orders of magnitude larger than the input, but we don't
maintain the output, we maintain the input.  I needed something quick
and dirty to recognize generated files and not include them in my LOC
metrics (unless requested).
.Pp
.Xr \fBmcloc 1\fR
was updated with relatively minor tweaks throughout the 2000's and
2010's.
.Pp
In 2019 and early 2020, I rewrote
.Xr \fBmcloc 1\fR
to use reentrant lexical analyzers and allow multiple threads (15
years overdue!).  I also restructured to use a two-level storage
scheme (directories and files within a directory) and
std::unordered_map.  The original std::map based implementation wasn't
terribly slow, but the way I used it (full paths as keys) meant longer
keys versus the two-level scheme and also hindered fast calculation of
LOC per directory.
.Pp
Since I'm now using std::filesystem instead of
.Xr \fBfts 3\fR ,
the code should be portable to Windows with a getopt() replacement
(I haven't tried yet).
.Pp
I also added the ability to show lines of code and comments per file
extension or per type, as well as the ability to sort by size.  In
addition, I added smarter code to recognize source code generated by
\fBflex(1)\fR or \fBbison(1)\fR; the current version recognizes
comments and preprocessor directives in the generated code instead of
relying on matching filenames.  The same is true for code generated
by Qt tools (moc, rcc, uic).  This allows me to ignore generated
files even if I can't find the generator source code in the given
directories, and doesn't require parsing flex and bison options in the
generator sources nor peeking at build infrastructure.  However,
\fBflex(1)\fR and \fBbison(1)\fR don't produce anything in their
output to indicate the source of the generated code if you disable
\fB#line\fR directives.  If you use the \fB-L\fR option of
\fBflex(1)\fR or the \fB-l\fR option of \fBbison(1)\fR, for example,
.Xr \fBmcloc 1\fR
will still recognize the generated code as generated, but will not be
able to name the source.
.Sh BUGS
.Xr \fBmcloc 1\fR
is driven primarily by filename extensions, by design.  That makes it
more suitable for code bases that have common source file naming conventions,
and unsuitable for code bases where files can't be easily identified
by file extension or simple regular expression matching.
.Pp
Prolog is disabled in the default configuration due to file extension
conflicts with Perl (.pl) and Qt project files ('.pro').
.Pp
The grouping of languages under scanners isn't particularly useful to
the end user.  It's an implementation detail which allows languages
with the same comment syntax to be handled by a single scanner.
.Sh AUTHORS
Daniel W. McRobb
.Mt dwm@mcplex.net
.\" I need to document the configuration!!!!


ChangeLog
2021-01-15  Daniel McRobb  <dwm@mcplex.net>

	* tagged as mcloc-1.0.6

	* Created separate Rust scanner so I can handle Rust nested
	comments.  Add a basic unit test for it using the comments from
	the Rust reference
	(https://doc.rust-lang.org/reference/comments.html).

2021-01-14  Daniel McRobb  <dwm@mcplex.net>

	* tagged as mcloc-1.0.5

2021-01-13  Daniel McRobb  <dwm@mcplex.net>

	* Updated the mcloc manpage.  Added a bit more documentation to
	the configuration file.

2021-01-12  Daniel McRobb  <dwm@mcplex.net>

	* Improved accuracy of Cpp and Flex scanners.

2021-01-09  Daniel McRobb  <dwm@mcplex.net>

	* Added Cmake scanner.  Added SQL scanner.

	* Added YAML scanning (no code change, just added to the Shell
	scanner configuration).

	* Added Assembly scanning (no code change, just added to Lisp
	scanner configuration).

	* Added IDL scanning (no code change, just added to Cpp scanner
	configuration).

	* Added protobuf scanning (no code change, just added to Cpp
	scanner configuration).

	* Added sed and awk scanning (no code change, just added to
	the Shell scanner configuration).

2021-01-07  Daniel W. McRobb  <dwm@mcplex.net>

	* Added Lua scanner, but it's not perfect.  State would have to be
	kept for the nesting form since the number of '=' between braces
	is indefinite but must be matched with the closing braces.
	']]' can appear in source, which makes block comments in
	Lua troublesome.  I've never used Lua, so I'll wait for some
	feedback.

	* Added Prolog scanner, but it's disabled in the default
	configuration file due to conflicts with Perl source file
	extensions and Qt project file extensions.

	* Added Fortran scanner.

	* Added Julia scanner.

	* Added Dart scanning.  No code changes, it uses the Cpp scanner.

	* Added Perl scanner.

	* Cleaned up some of the build infrastructure.

	* tagged as mcloc-1.0.4

2021-01-06  Daniel McRobb  <dwm@mcplex.net>

	* First pass at a GUI application is done (apps/qmcloc).

2021-01-04  Daniel McRobb  <dwm@mcplex.net>

	* Added scanners for TeX/LaTeX and HTML.

2021-01-03  Daniel W. McRobb  <dwm@mcplex.net>

	* apps/mcloc: Moved command line processing classes to Mcloc
	namespace to avoid any potential future conflict with same classes
	in libDwm.

2021-01-02  Daniel W. McRobb  <dwm@mcplex.net>

	* Reorganized to build a library so I can later build other
	applications.  mcloc is now in the apps/mcloc directory,
	classes that can be used by other applications are in the classes
	directory.

2020-12-23  Daniel W. McRobb  <dwm@mcplex.net>

	* Makefile: Use mkdebcontrol (from libDwm), dpkg-deb and dpkg-name
	instead of epm when building debian packages on linux.

2020-11-07  Daniel McRobb  <dwm@mcplex.net>

	* tagged as mcloc-1.0.3

	* mcloc.cc: Added '-v' command line option to show version.

	* tagged as mcloc-1.0.2

	* mcloc.cc:
	If std::thread::hardware_concurrency() returns a value greater
	than 24, reduce it by a factor of 2 before using it as the default
	number of threads.  On the platforms I own with 24 or more cores,
	it's much faster.  This appears to be due to I/O thrashing, which
	makes some sense.  It makes a glaring difference on Threadripper
	3960X with PCIe Gen4 x4 NVMe storage (Sabrent Rocket 1TB); 60
	million lines/sec versus 30 million lines/sec.

	* DwmMclocSourceCollection.cc:
	Fix bug with calculation of trailer string length that I use for
	full-terminal-width underlines; if the line is aleady wide, the
	trailer should just be empty.

2020-11-06  Daniel McRobb  <dwm@mcplex.net>

	* DwmMclocCodeUtils.cc:
	Explicitly ignore shared library files, i.e. those with filenames
	matching the ECMAScript regular expression ".+\.so(\.[0-9]+)*".


apps/mcloc/mcloc.cc
//===========================================================================
// @(#) $DwmPath: dwm/mcplex/mcloc/trunk/apps/mcloc/mcloc.cc 11834 $
//===========================================================================
//  Copyright (c) Daniel W. McRobb 1998, 2000, 2004, 2012, 2019-2021
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file mcloc.cc
//!  \brief count lines of code in file(s) and/or directories
//---------------------------------------------------------------------------

#include <filesystem>
#include <iostream>
extern "C" {
  #include <unistd.h>
}

#include <cstdlib>
#include <regex>

#include "DwmMclocArguments.hh"
#include "DwmMclocConfig.hh"
#include "DwmMclocSourceCollection.hh"
#include "DwmMclocFileScanner.hh"
#include "DwmMclocPackageInfo.hh"
#include "DwmMclocCocomo1Intermediate.hh"

using namespace std;

static const string  g_defaultConf("/usr/local/etc/mcloc.cfg");

static const string
g_defaultCocomoConf("/usr/local/etc/mcloc_cocomo1_intermediate.cfg");

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static set<string> & LowerStrings(set<string> & ss)
{
  for (const auto & str : ss) {
    ss.erase(str);
    ss.insert(Dwm::Mcloc::CodeUtils::tolower(str));
  }
  return ss;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static string SupportedLanguages()
{
  using namespace Dwm::Mcloc;

  string          rc;
  vector<string>  languages;
  Config::Config  config;
  if (config.Parse(g_defaultConf)) {
    string  sep;
    for (const auto & scanner : config.Scanners()) {
      for (const auto & language : scanner.Languages()) {
        languages.push_back(language.Name());
      }
    }
    sort(languages.begin(), languages.end(),
         [] (const string & s1, const string & s2)
         {
           return (Dwm::Mcloc::CodeUtils::tolower(s1) < Dwm::Mcloc::CodeUtils::tolower(s2));
         });
    for (const auto & lang : languages) {
      rc += sep + lang;
      sep = ", ";
    }
  }
  return rc;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static string SupportedScanners()
{
  string    scanners;
  Dwm::Mcloc::Config::Config  config;
  if (config.Parse(g_defaultConf)) {
    string  sep;
    for (const auto & scanner : config.Scanners()) {
      scanners += sep + scanner.Name();
      sep = ", ";
    }
  }
  return scanners;
  
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
typedef   Dwm::Mcloc::Arguments<Dwm::Mcloc::Argument<'C',bool>,
                                Dwm::Mcloc::Argument<'E',set<string>>,
                                Dwm::Mcloc::Argument<'L',set<string>>,
                                Dwm::Mcloc::Argument<'S',set<string>>,
                                Dwm::Mcloc::Argument<'c',string>,
                                Dwm::Mcloc::Argument<'d',bool>,
                                Dwm::Mcloc::Argument<'e',bool>,
                                Dwm::Mcloc::Argument<'f',bool>,
                                Dwm::Mcloc::Argument<'l',bool>,
                                Dwm::Mcloc::Argument<'s',bool>,
                                Dwm::Mcloc::Argument<'g',bool>,
                                Dwm::Mcloc::Argument<'j',int>,
                                Dwm::Mcloc::Argument<'o',bool>,
                                Dwm::Mcloc::Argument<'q',bool>,
                                Dwm::Mcloc::Argument<'a',bool>,
                                Dwm::Mcloc::Argument<'r',bool>,
                                Dwm::Mcloc::Argument<'m',bool>,
                                Dwm::Mcloc::Argument<'M',string>,
                                Dwm::Mcloc::Argument<'v',bool>> MyArgType;

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void InitArgs(MyArgType & args, const std::string & configPath)
{
  args.SetHelp<'C'>("print the configuration");
  args.SetValueName<'E'>("extensions");
  args.SetHelp<'E'>("only scan files with extensions in the given"
                    " comma-separated list of extensions");
  args.SetValueName<'L'>("languages");
  args.SetHelp<'L'>("only scan files for the languages in the given"
                    " comma-separated list of languages, which may"
                    " include:\n\n" + SupportedLanguages() );
  args.SetValueName<'S'>("scanners");
  args.SetHelp<'S'>("only enable scanners in the given comma-separated"
                    " list of scanners, which may include:\n\n"
                    + SupportedScanners());
  args.Set<'c'>(configPath);
  args.SetValueName<'c'>("config");
  args.SetHelp<'c'>("specify configuration file");
  args.SetHelp<'d'>("print directory counters");
  args.SetHelp<'e'>("print counters by file extension");
  args.SetHelp<'f'>("print counters by file (default)");
  args.SetHelp<'l'>("print counters by language");
  args.SetHelp<'s'>("print counters by scanner type");
  args.SetHelp<'g'>("include generated files");
  args.Set<'j'>(thread::hardware_concurrency());
  if (args.Get<'j'>() > 24) {
    args.Set<'j'>(args.Get<'j'>() / 2);
  }
  args.SetValueName<'j'>("numThreads");
  args.SetHelp<'j'>("specify the number of parsing threads (default "
                    + to_string(args.Get<'j'>()) + ")");
  args.SetHelp<'o'>("do not recurse into subdirectories");
  args.SetHelp<'q'>("only show total");
  args.SetHelp<'a'>("sort output by size, ascending");
  args.SetHelp<'r'>("sort output by size, descending");
  args.SetHelp<'m'>("show COCOMO estimates");
  args.SetHelp<'M'>("specify COCOMO configuration file");
  args.SetHelp<'v'>("show version and exit");
  
  args.SetConflicts({ {'d','e','f','l','s'}, {'a','r'} });

  args.LoadFromEnvironment("MCLOC");
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void
PrintCocomoEstimates(const Dwm::Mcloc::SourceCollection & sc,
                     const string & cocoCfgPath)
{
  namespace Intermediate = Dwm::Mcloc::Cocomo1::Intermediate;
  
  Intermediate::Config  cocoCfg;
  if (cocoCfg.Parse(cocoCfgPath)) {
    Dwm::Mcloc::CodeCounter  totals;
    sc.GetTotal(totals);
    double  effort = Intermediate::Effort(cocoCfg, totals.CodeLines());
    double  devTime =
      Intermediate::DevelopmentTime(cocoCfg, totals.CodeLines());
    double  persons = effort / devTime;

    Dwm::Mcloc::TerminalTricks  tt;
    if (isatty(STDOUT_FILENO)) {
      cout.imbue(std::locale(""));
    }
    else {
      tt.On(false);
    }
    tt.Underscore(cout, true);
    tt.Bold(cout, true);
    cout << '\n'
         << setw(22) << "Effort (person-months)";
    tt.Underscore(cout, false);
    cout << "    ";
    tt.Underscore(cout, true);
    cout << setw(13) << "Time (months)";
    tt.Underscore(cout, false);
    cout << "    ";
    tt.Underscore(cout, true);
    cout << setw(6) << "People" << '\n';
    tt.Underscore(cout, false);
    tt.Bold(cout, false);
    cout << setw(22) << setprecision(4) << effort << "    "
         << setw(13) << setprecision(4) << devTime << "    "
         << setw(6) << setprecision(4) << persons << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  string  configPath(g_defaultConf);
  string  homeConfig(string(getenv("HOME")) + "/.mcloc");
  if (filesystem::exists(homeConfig)
      && filesystem::is_regular_file(homeConfig)) {
    //  We have a personal configuration file in our home directory.
    configPath = homeConfig;
  }

  using Dwm::Mcloc::SourceCollection;
    
  MyArgType  args;
  InitArgs(args, configPath);
  
  int  argind = args.Parse(argc, argv);
  if (argind < 0) {
    cerr << args.Usage(argv[0], "path(s)");
    exit(1);
  }

  if (args.Get<'v'>()) {
    cout << Dwm::Mcloc::PackageInfo::Name() << '-'
         << Dwm::Mcloc::PackageInfo::Version() << '\n';
    exit(0);
  }
  
  int  numThreads = args.Get<'j'>();
  
  Dwm::Mcloc::Config::Config  config;
  if (config.Parse(configPath)) {
    if (! args.Get<'E'>().empty()) {
      config.KeepOnlyExtensions(args.Get<'E'>());
    }
    if (! args.Get<'L'>().empty()) {
      config.KeepOnlyLanguages(LowerStrings(args.Get<'L'>()));
    }
    if (! args.Get<'S'>().empty()) {
      config.KeepOnlyScanners(LowerStrings(args.Get<'S'>()));
    }
    
    if (args.Get<'C'>()) {
      cout << configPath;
      if (! args.Get<'E'>().empty()) { cout << ", modified by -E"; }
      if (! args.Get<'L'>().empty()) { cout << ", modified by -L"; }
      if (! args.Get<'S'>().empty()) { cout << ", modified by -S"; }
      cout << ":\n\n" << config;
      return 0;
    }
    
    if (argind >= argc) {
      cerr << args.Usage(argv[0], "path(s)");
      exit(1);
    }

    Dwm::Mcloc::CodeUtils::Initialize(config);

    Dwm::Mcloc::HowToSort  howToSort = Dwm::Mcloc::sortByName;
    if (args.Get<'a'>())      { howToSort = Dwm::Mcloc::sortBySize; }
    else if (args.Get<'r'>()) { howToSort = Dwm::Mcloc::sortBySizeReversed; }

    SourceCollection::WhatToPrint  printBy = SourceCollection::byFile;
    if (args.Get<'l'>()) { printBy = SourceCollection::byLanguage; }
    else if (args.Get<'d'>()) { printBy = SourceCollection::byDirectory; }
    else if (args.Get<'e'>()) { printBy = SourceCollection::byExtension; }
    else if (args.Get<'f'>()) { printBy = SourceCollection::byFile; }
    else if (args.Get<'s'>()) { printBy = SourceCollection::byScanner; }

    SourceCollection  sc;
    for (int i = argind; i < argc; ++i) {
      if (filesystem::exists(argv[i])) {
        if (filesystem::is_directory(argv[i])) {
          sc.AddDirectory(argv[i], (! args.Get<'o'>()));
        }
        else if (filesystem::is_regular_file(argv[i])) {
          sc.AddFile(argv[i]);
        }
      }
    }
    sc.Parse(numThreads);
    cout.imbue(std::locale(""));
    
    sc.Print(cout, printBy, args.Get<'g'>(), howToSort, args.Get<'q'>());

    if (args.Get<'m'>()) {
      string  cocoCfgPath = g_defaultCocomoConf;
      if (! args.Get<'M'>().empty()) {
        cocoCfgPath = args.Get<'M'>();
      }
      PrintCocomoEstimates(sc, cocoCfgPath);
    }
    return 0;
  }
  else {
    cerr << "Failed to parse configuration in '" << configPath << "'\n";
    args.Usage(argv[0], "path(s)");
  }
  return 1;
}
